scratch/augmented_lagrangian.py

```diff
@@ -0,0 +1,79 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+#
+# Implements the augmented Lagrangian method for constrained
+# nonlinear least squares as described on page 11.21 of
+# http://www.seas.ucla.edu/~vandenbe/133B/lectures/nllseq.pdf
+
+import copy
+import torch
+import theseus as th
+
+
+def solve_augmented_lagrangian(
+    err_fn,
+    optim_vars,
+    aux_vars,
+    dim,
+    equality_constraint_fn,
+    optimizer_cls,
+    optimizer_kwargs,
+    verbose,
+    initial_state_dict,
+    num_augmented_lagrangian_iterations=10,
+    callback=None,
+):
+    def err_fn_augmented(optim_vars, aux_vars):
+        # TODO: Dangerous to override optim_vars,aux_vars here
+        original_aux_vars = aux_vars[:-2]
+        original_err = err_fn(optim_vars, original_aux_vars)
+        g = equality_constraint_fn(optim_vars, original_aux_vars)
+
+        mu, z = aux_vars[-2:]
+        sqrt_mu = torch.sqrt(mu.tensor)
+        err_augmented = sqrt_mu * g + z.tensor / (2.0 * sqrt_mu)
+        combined_err = torch.cat([original_err, err_augmented], axis=-1)
+        return combined_err
+
+    g = equality_constraint_fn(optim_vars, aux_vars)
```

Comment:
For this, I like the `RobustCost`-style wrapper, as you mentioned in the TODOs. Something like
```python
equality_constraint = th.EqualityConstraint(th.AutoDiffCostFunction(...))
```
Following on the other comment above, a class like `AugmentedLagrangian` could then iterate over all cost functions in the objective, and add extra cost terms for any costs of type `EqualityConstraint` that computes the `err_augmented` term corresponding to each equality constraint.